src/coreclr/jit/promotion.cpp

```diff
@@ -0,0 +1,1310 @@
+    //
+    void StoreBeforeReturn(GenTreeUnOp* ret)
+    {
+        if (ret->TypeIs(TYP_VOID) || !ret->gtGetOp1()->OperIs(GT_LCL_VAR, GT_LCL_FLD))
+        {
+            return;
+        }
+
+        GenTreeLclVarCommon* retLcl = ret->gtGetOp1()->AsLclVarCommon();
+        if (retLcl->TypeIs(TYP_STRUCT))
+        {
+            unsigned size = retLcl->GetLayout(m_compiler)->GetSize();
+            WriteBackBefore(&ret->gtOp1, retLcl->GetLclNum(), retLcl->GetLclOffs(), size);
+        }
+    }
+
+    //------------------------------------------------------------------------
+    // WriteBackBefore:
+    //   Update the use with IR that writes back all necessary overlapping
+    //   replacements into a struct local.
+    //
+    // Parameters:
+    //   use  - The use, which will be updated with a cascasing comma trees of assignments
+    //   lcl  - The struct local
+    //   offs - The starting offset into the struct local of the overlapping range to write back to
+    //   size - The size of the overlapping range
+    //
+    void WriteBackBefore(GenTree** use, unsigned lcl, unsigned offs, unsigned size)
+    {
+        if (m_replacements[lcl] == nullptr)
+        {
+            return;
+        }
+
+        jitstd::vector<Replacement>& replacements = *m_replacements[lcl];
+        size_t                       index        = BinarySearch<Replacement, &Replacement::Offset>(replacements, offs);
+
+        if ((ssize_t)index < 0)
+        {
+            index = ~index;
+            if ((index > 0) && replacements[index - 1].Overlaps(offs, size))
+            {
+                index--;
+            }
+        }
+
+        unsigned end = offs + size;
+        while ((index < replacements.size()) && (replacements[index].Offset < end))
+        {
+            Replacement& rep = replacements[index];
+            if (rep.NeedsWriteBack)
+            {
+                GenTree*   dst = m_compiler->gtNewLclFldNode(lcl, rep.AccessType, rep.Offset);
+                GenTree*   src = m_compiler->gtNewLclvNode(rep.LclNum, rep.AccessType);
+                GenTreeOp* comma =
+                    m_compiler->gtNewOperNode(GT_COMMA, (*use)->TypeGet(), m_compiler->gtNewAssignNode(dst, src), *use);
+                *use = comma;
+                use  = &comma->gtOp2;
+
+                rep.NeedsWriteBack = false;
+                m_madeChanges      = true;
+            }
+
+            index++;
+        }
+    }
+
+    //------------------------------------------------------------------------
+    // MarkForReadBack:
+    //   Mark that replacements in the specified struct locals need to be read
+    //   back before their next use.
+    //
+    // Parameters:
+    //   lcl          - The struct local
+    //   offs         - The starting offset of the range in the struct local that needs to be read back from.
+    //   size         - The size of the range
+    //   conservative - Whether this is a potentially conservative read back
```

Comment:
Is this for the case where you may be able to prove that a reassembled struct won't be modified?

Reply:
This is to make it clear in the jitdump that this is an assignment. Eventually we want to decompose these directly (like block morphing) and avoid this read back in many cases, but the logic is not implemented yet.